Add MeasurementProcessor to Metrics SDK (#4318)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

|     ...          | Measurements... |                      | Measurements... |                 |
|     ...          |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     ...          |                 |                      |                 |                 |
|     ...          |                 +----------------------+                 +-----------------+
+------------------+
```

### MeasurementProcessor operations

#### OnMeasure

`OnMeasure` is called when a `Measurement` is recorded. This method is called synchronously on the thread that emitted the `Measurement`, therefore it SHOULD NOT block or throw exceptions.

**Parameters:**

* `measurement` - a [Measurement](./api.md#measurement) that was recorded
* `context` - the resolved `Context` (the explicitly passed `Context` or the current `Context`)

**Returns:** Void

For a `MeasurementProcessor` registered directly on SDK `MeterProvider`, the `measurement` mutations MUST be visible in next registered processors.

A `MeasuremenetProcessor` may freely modify `measurement` for the duration of the `OnMeasure` call.

> A `MeasurementProcessor` MUST invoke `OnMeasure` on the next registered processor.

Reviewer: How would a MeasurementProcessor do this? Is there a mechanism for the processor to get the "next registered processor"?

### Built-in processors

The standard OpenTelemetry SDK MUST implement simple processor as described below.

#### SimpleProcessor

> This is an implementation of `MeasurementProcessor` which calculates an in-memory state from incoming `Measurements`.

Reviewer: Sounds like the intention is to have a "DefaultProcessor" rather than emphasizing "Simple"?

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## MeasurementProcessor

**Status**: [Development](../document-status.md)

`MeasurementProcessor` is an interface which allows hooks when a `Measurement` is recorded by an `Instrument`.

Built-in measurement processors are responsible for [Measurement Processing](#measurement-processing).

`MeasurementProcessors` can be registered directly on SDK `MeterProvider` and they are invoked in the same order as they were registered.

> Each processor registered on the `MeterProvider` is part of a pipeline.

Reviewer 2: Should note that the pipeline always concludes with built in DefaultProcessor.

Author: I updated this in [e6eec36](https://github.com/open-telemetry/opentelemetry-specification/pull/4318/commits/e6eec36d0207562123b1c529b33bd7c974474fec).

CC @jack-berg @cijothomas 

The alternative was to have SDK implement and ensure that a `NoopProcessor` (removed [here](https://github.com/open-telemetry/opentelemetry-specification/pull/4318/commits/f57fe00625586e9438b972943a2a407fc01bcd13)) is always the last in the chain. The `NoopProcessor`, by definition, does not call `next()`. This would align with other SDKs where users have to explicitly add a processor that connects the pipeline to the exporter - otherwise, they wouldn't see any data.

My only argument against making the `SDK` ensure that the pipeline ends with the `DefaultProcessor` is that removing this requirement will be difficult going forward.

Reviewer 3: Could this be implicit and not something the user has to configure? I'm wondering if we need to expose the DefaultProcessor at all.

Author: Yes, the current spec says:

```
SDK MUST ensure that the pipeline concludes with the built-in [DefaultProcessor](#defaultprocessor).
```

It says nothing about making `DefaultProcessor` public - it very well can be hidden from the user. If SDK adds it automatically, I don't see a reason for exposing it and breaking the encapsulation.

Reviewer 3: Since it is called out as a top-level thing in this PR, I assumed it is public.

|     Instrument X |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     Instrument Y +                 |                      |                 |                 |
|   Meter B        |                 +----------------------+                 +-----------------+
|     Instrument Z |
|     ...          |                 +----------------------+                 +-----------------+
|     ...          | Measurements... |                      | Measurements... |                 |
|     ...          |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     ...          |                 |                      |                 |                 |
|     ...          |                 +----------------------+                 +-----------------+
+------------------+
```

### MeasurementProcessor operations

#### OnMeasure

`OnMeasure` is called when a `Measurement` is recorded. This method is called synchronously on the thread that emitted the `Measurement`, therefore it SHOULD NOT block or throw exceptions.

**Parameters:**

* `context` - the resolved `Context` (the explicitly passed `Context` or the current `Context`)

> * `measurement` - a [Measurement](./api.md#measurement) that was recorded

Reviewer 2: I'm planning on exposing all the components of a measurement (i.e. value, attributes, instrument descriptor) as individual parameters, since in java, bundling them up in a measurement carrier would require allocating measurement. I don't think anything needs to be done in this PR. Just mentioning. 

I will note that the API definition of measurement only includes value and attributes, so as it stands, there's not way for a processor to perform logic based on the instrument. I think the conclusion of this [conversation](https://github.com/open-telemetry/opentelemetry-specification/pull/4318/files#r1925726339) was that we wanted to include a parameter which encapsulates instrument metadata (name, description, unit, type).

> * `next` - a callback to invoke `OnMeasure` on the next processor in the chain. It MUST be callable without a reference to the next processor.

Reviewer: I feel this might be too restrictive; we might want to give more flexibility here so different programming languages can find their own idiomatic way or performant way.

How about this?

```suggestion
* `next` - this allows the MeasurementProcessor to pass the measurements to the next MeasurementProcessor in the chain. It can be a reference to the next MeasurementProcessor, a bound callback to invoke `OnMeasure` on the next processor in the chain without an explicit reference to the next processor, or something else. [OpenTelemetry SDK](../overview.md#sdk) authors MAY decide the language idiomatic approach.
```

**Returns:** Void

For a `MeasurementProcessor` registered directly on SDK `MeterProvider`, the `measurement` mutations MUST be visible in next registered processors.

A `MeasuremenetProcessor` MAY freely modify `measurement` for the duration of the `OnMeasure` call.

A `MeasurementProcessor` SHOULD invoke `next`. A `MeasurementProcessor` MAY decide to drop the `Measurement` by not invoking the next processor.

### Built-in processors

The standard OpenTelemetry SDK MUST implement default processor as described below.

#### DefaultProcessor

> This is an implementation of `MeasurementProcessor` which calculates an in-memory state from incoming `Measurements`.

Reviewer 4: should the SDK ensure that this is the last processor in the chain?

Author: I didn't specify this to keep it in line with other Logs and Traces SDKs. Do we have good argument to make SDK ensure this processor is included in the pipeline?

Reviewer 4: if this is not included in the pipeline (and ensure it is the last one), how does MeasurementProcessors work ?
For example: User adds a measurement processor. In the OnMeasure(), it will receive a handle to invoke the next. if the user calls next.OnMeasure(), aggregation/default is expected to occur. But without having the default processor already there, this will not work right? 
Also, the DefaultProcessor should be the last, otherwise, if user's intention was to "drop" the measurement, it will be pointless if DefaultProcessor is not the last one.

Reviewer 4: For Logs,Traces, the processor was not responsible for invoking the next processor. But the MeasurementProcessor proposed here is different in the sense it is responsible for calling the next (or not calling) processor. This is why I asked who is responsible for ensuring that the last MeasurementProcessor is the DefaultProcessor.

Author: Moving the conversation to https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r1934793288.

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## MeasurementProcessor

**Status**: [Development](../document-status.md)

`MeasurementProcessor` is an interface which allows hooks when a `Measurement` is recorded by an `Instrument`.

Built-in measurement processors are responsible for [Measurement Processing](#measurement-processing).

`MeasurementProcessors` can be registered directly on SDK `MeterProvider` and they are invoked in the same order as they were registered.

Each processor registered on the `MeterProvider` is part of a pipeline.

SDK MUST allow users to implement and configure custom processors.

> SDK MUST ensure that the pipeline concludes with the built-in [DefaultProcessor](#defaultprocessor).

Reviewer 5: This conflicts with

> A `MeasurementProcessor` SHOULD invoke `next`. A `MeasurementProcessor` MAY decide to drop the `Measurement` by not invoking the next processor.

I think we should say something like the SDK pass DefaultProcessor to the chain. 

I am not sure if with the current design the SDK would  know how/when to pass the DefaultProcessor to the processing chain.

I have a feeling that we would need something like a http.Handler middleware pattern popular in Go https://pkg.go.dev/github.com/gorilla/mux#MiddlewareFunc. The DefaultProcessor does not need to be exposed by the SDK. Hard to validate without a prototype 😉

Reviewer 5: Discussed offline. The design of Go HTTP middlewares is different as it is primary goal to compose nicely with `http.Handler` (which is like a sink interface which we do not provide). The  design sketched out by @Blinkuu https://go.dev/play/p/wPZRm5xk3nO looks nice.

|   Meter A        | Measurements... |                      | Measurements... |                 |
|     Instrument X |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     Instrument Y +                 |                      |                 |                 |
|   Meter B        |                 +----------------------+                 +-----------------+
|     Instrument Z |
|     ...          |                 +----------------------+                 +-----------------+
|     ...          | Measurements... |                      | Measurements... |                 |
|     ...          |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     ...          |                 |                      |                 |                 |
|     ...          |                 +----------------------+                 +-----------------+
+------------------+
```

### MeasurementProcessor operations

#### OnMeasure

`OnMeasure` is called when a `Measurement` is recorded. This method is called synchronously on the thread that emitted the `Measurement`, therefore it SHOULD NOT block or throw exceptions.

> **Parameters:**

Reviewer 5: Shouldn't  the `MeasurementProcessor` have also access to instrumentation scope (connected with the instrument) and resource (used by the provider)? The LogRecord and Span processors have access to this data.

Reviewer 5: > The primary reason is we don't want to impose implementation details, allowing for optimal/idiomatic approaches.

The languages do not need to copy/implement the abstractions one to one. 

> The high-level idea is that the Measurement type should be defined as a 3-tuple (Value, Attributes, Instrument).

I totally agree with it and I feel that it must be defined. Otherwise languages may not support all of the data which is relevant for processing.

Author: This is the reason it's not explicit: https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r1934789623

Reviewer 5: I think we should explicitly call out them as separate parameters in the specification. Then it is the language SIG decision if they want to combine them into individual types or keep them separated. I think this would be fine for both @jack-berg (https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r1934789623) and me (https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r1963454443).

@jack-berg Does it sound reasonable?

@Blinkuu Please do not make any changes until at least @jack-berg agrees :wink:

Reviewer 2: > @jack-berg Does it sound reasonable?

Sorry I obviously haven't looked at this in a while. I'm not opinionated on how we accomplish it, but I agree with @pellared that we'll want to make scope and resource available.

Reviewer 2: Maybe worth saying that implementations should ensure that add additional resource, scope, and instrument metadata should be accessible to `onMeasure`

Author: @jack-berg Do you expect any changes, or should we defer it and/or update the `Measurement` definition to include instrument metadata?

* `context` - the resolved `Context` (the explicitly passed `Context` or the current `Context`)
* `measurement` - a [Measurement](./api.md#measurement) that was recorded

> * `next` - this allows the `MeasurementProcessor` to pass the measurements to the next `MeasurementProcessor` in the chain. It can be a reference to the next `MeasurementProcessor`, a bound callback to invoke `OnMeasure` on the next processor in the chain without an explicit reference to the next processor, or something else. [OpenTelemetry SDK](../overview.md#sdk) authors MAY decide the language idiomatic approach.

Reviewer 3: Not sure how liberally we can interpret the "language idiomatic approach". For example, would it be reasonable to allow `OnMeasure()` return a new Measurement and have the SDK chain them together instead of recursing?

```python
for mp in mps:
  measurement = mp.on_measure(measurement)
  if measurement == "DROP":
    break
```

Author: The current design with `next()` allows duplicating measurements into downstream processors. For example, you can take one measurement and transform it into two. We discussed an API like this where the `OnMeasure()` function would return a list of measurements. The conclusion was that such an API would be slightly confusing.

Reviewer 5: > I see some people arguing we shouldn't allow duplication of measurements

Can you link to the comments? Are there any reasons why it should be disallowed? If there is no good reason I think that the design should be open to different advanced use cases (e.g. like this for logs: https://github.com/open-telemetry/opentelemetry-specification/pull/4407).

Author: @pellared Please see https://github.com/open-telemetry/opentelemetry-specification/pull/4318/files#r1925726339

> * `next` - this allows the `MeasurementProcessor` to pass the measurements to the next `MeasurementProcessor` in the chain. It can be a reference to the next `MeasurementProcessor`, a bound callback to invoke `OnMeasure` on the next processor in the chain without an explicit reference to the next processor, or something else. [OpenTelemetry SDK](../overview.md#sdk) authors MAY decide the language idiomatic approach.

Reviewer 3: May a processor create new measurements by invoking `next()` multiple times? Dumb example

```py
def on_measure(measurement, next)
  measurement_squared = measurement.copy()
  measurement_squared.attributes["is_squared"] = True
  measurement_squared.value = measurement2.value**2
  
  next(measurement)
  next(measurement_squared)
```

Author: Yes, this is intentional in design to cover some custom use cases, like mapping a measurement into two measurements of different instrument types.

Author: Yes, this is currently possible. I see some conflicting views regarding this. The use cases I wanted to cover do not require it, but it's been raised in this PR before. I'm happy to revisit this.

Reviewer 5: I think it makes sense, but similarly to e.g. logs we should document what a measurement processor can mutate when dealing with measurements. For instance, I think that instrumentation scope and resource should come as arguments to `OnMeasure` but `next` should not accept neither instrumentation scope and resource. This way the processor could not be able to change the instrumentation scope nor resource (similarly to LogRecordProcessor and SpanProcessor)

**Returns:** Void

For a `MeasurementProcessor` registered directly on SDK `MeterProvider`, the `measurement` mutations MUST be visible in next registered processors.

> A `MeasuremenetProcessor` MAY freely modify `measurement` for the duration of the `OnMeasure` call.

Reviewer 3: ```suggestion
A `MeasurementProcessor` MAY freely modify `measurement` for the duration of the `OnMeasure` call.
```

> A `MeasurementProcessor` SHOULD invoke `next`. A `MeasurementProcessor` MAY decide to drop the `Measurement` by not invoking the next processor.

Reviewer 3: I imagine some possible use cases where users would expect the whole pipeline to run regardless. For example, a self observability processor that counts the number of dropped measurements.

Author: > I imagine some possible use cases where users would expect the whole pipeline to run regardless.
I'm not following. Run regardless of what?

Reviewer 3: Run regardless of if the point is dropped

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## MeasurementProcessor

**Status**: [Development](../document-status.md)

> `MeasurementProcessor` is an interface which allows hooks when a `Measurement` is recorded by an `Instrument`.

Reviewer 6: In the context of tracing, a [processor](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-processor) is somewhat of a different thing (it's not a filterer of data, it sits on top of exporters).

Should we name this differently than `Processor` to differenciate the two?
Or should we make both behaviors similar?

Author: > In the context of tracing, a [processor](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-processor) is somewhat of a different thing (it's not a filterer of data, it sits on top of exporters).

The `MeasurementProcessor` is more than a filter of data. It can modify the measurement, decide to drop it, or even emit it multiple times.

> Should we name this differently than Processor to differentiate the two?
Or should we make both behaviors similar?

To me, they are semantically similar. The difference is in the architecture of the Metrics SDK, which makes it hard to fit and connect the new processor pipeline directly with exporters. We could revisit this in the future and try to align the behavior between Logs, Traces, and Metrics SDKs. To keep an open door for such ideas, I'm purposefully specifying that:

`Built-in measurement processors are responsible for Measurement Processing.`

Where `Measurement Processing` is currently _vaguely_ defined in this spec by using the word `SHOULD`.

Reviewer 7: I support @Blinkuu's reasoning above. In my thinking, we can use "processor" across OpenTelemetry when a component in a program receives and outputs the same type of data. Trace SDK processors input/output span data. Collector processors input/output pipeline data. Things that are not processors (e.g., Metric reader, Trace sampler, Exporters, Receivers) generally have different types of input and output.

Here, MeasurementProcessor is appropriate because the input and the output types are the same.

Reviewer: > Here, MeasurementProcessor is appropriate because the input and the output types are the same.

+1

Built-in measurement processors are responsible for [Measurement Processing](#measurement-processing).

> `MeasurementProcessors` can be registered directly on SDK `MeterProvider` and they are invoked in the same order as they were registered.

Reviewer 4: > they are invoked in the same order as they were registered.

Unclear how is this part works with the later wording "_The SDK MUST inject the `OnMeasure` function from the next processor in the chain into the current processor_"

Who is invoking the MeasurementProcessors in the order? Is that MeterProvider/SDK ? Or is it the responsibility of each `MeasurementProcessor` to invoke the next one in the chain?

Author: > Unclear how is this part works with the later wording "The SDK MUST inject the OnMeasure function from the next processor in the chain into the current processor"
>
> Who is invoking the MeasurementProcessors in the order? Is that MeterProvider/SDK ? Or is it the responsibility of each MeasurementProcessor to invoke the next one in the chain?

I have changed the wording around the `next` parameter, as the previous spec was confusing. The `next` parameter is a callback that MUST NOT require a reference to the next processor in the chain. On the implementation level, it will most likely end up as a closure. I see it as the SDK's responsibility to ensure processors are executed in the order they were configured on the `MeterProvider`.

That being said, every processor is able to decide whether it wants to invoke a chain of processors that follow it (by calling `next(context, measurement)`). Each processor can decide to call `next` either `0` (drop measurement), `1` (default flow), or `N` (custom use cases) times.

> Each processor registered on the `MeterProvider` is part of a pipeline. The SDK MUST inject the `OnMeasure` function from the next processor in the chain into the current processor. It is RECOMMENDED that the SDK adds a `NoopProcessor` at the end of a pipeline.

Reviewer 6: > It is RECOMMENDED that the SDK adds a `NoopProcessor` at the end of a pipeline.

What is the purpose of that? How would that be different from the SDK knowing a processor is the last one and stopping its processing? (this phrase feels like an implementation detail).

|   Meter A        | Measurements... |                      | Measurements... |                 |
|     Instrument X |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     Instrument Y +                 |                      |                 |                 |
|   Meter B        |                 +----------------------+                 +-----------------+
|     Instrument Z |
|     ...          |                 +----------------------+                 +-----------------+
|     ...          | Measurements... |                      | Measurements... |                 |
|     ...          |-----------------> MeasurementProcessor +-----------------> In-memory state |
|     ...          |                 |                      |                 |                 |
|     ...          |                 +----------------------+                 +-----------------+
+------------------+
```

### MeasurementProcessor operations

#### OnMeasure

`OnMeasure` is called when a `Measurement` is recorded. This method is called synchronously on the thread that emitted the `Measurement`, therefore it SHOULD NOT block or throw exceptions.

**Parameters:**

> * `context` - the resolved `Context` (the explicitly passed `Context` or the current `Context`)

Reviewer: Do we want `context` to be the 1st parameter? I think we want it to be the last one.

* `measurement` - a [Measurement](./api.md#measurement) that was recorded

> * `next` - the `OnMeasure` function from the next processor in the chain

Reviewer: Is `next` the "next processor" or "the OnMeasure function of the next processor"?

Author: See https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r1925111998

**Returns:** Void

For a `MeasurementProcessor` registered directly on SDK `MeterProvider`, the `measurement` mutations MUST be visible in next registered processors.

A `MeasuremenetProcessor` MAY freely modify `measurement` for the duration of the `OnMeasure` call.

> A `MeasurementProcessor` SHOULD invoke `OnMeasure` on the next registered processor. A `MeasurementProcessor` MAY decide to drop the `Measurement` by not invoking the next processor.

Reviewer: How would a MeasurementProcessor do that? For example, if it needs to invoke OnMeasure (which requires "next"), how would it figure out all the arguments?

### Built-in processors

The standard OpenTelemetry SDK MUST implement default processor as described below.

#### DefaultProcessor

This is an implementation of `MeasurementProcessor` which calculates an in-memory state from incoming `Measurements`.

#### NoopProcessor

> This is an implementation of `MeasurementProcessor` which does nothing. It can be used to entirely disable processing of `Measurements` or to safely terminate a chain of processors.

Reviewer 4: why is this required? View already allow users to disable any individual instrument, so is there a need for this NoopProcessor?

Author: Agreed that we can leave this out of specification. I've removed the recommendation to implement `NoopProcessor`.

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## MeasurementProcessor

> 

Reviewer 5: ```suggestion

**Status**: [Development](../document-status.md)

```

In particular, if any `MeasurementProcessor` has any associated exporter, it SHOULD try to call the exporter's `Export` with all `Measurements` for which this was not already done and then invoke `ForceFlush` on it. If a timeout is specified (see below), the `MeasurementProcessor` MUST prioritize honoring the timeout over finishing all calls. It MAY skip or abort some or all `Export` or `ForceFlush` calls it has made to achieve this goal.

`ForceFlush` SHOULD provide a way to let the caller know whether it succeeded, failed or timed out.

`ForceFlush` SHOULD only be called in cases where it is absolutely necessary, such as when using some FaaS providers that may suspend the process after an invocation, but before the `MeasurementProcessor` exports the emitted `Measuremenets`.

`ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` can be implemented as a blocking API or an asynchronous API which notifies the caller via a callback or an event. OpenTelemetry SDK authors can decide if they want to make the flush timeout configurable.

> 

Author: The big question is whether we want to make `MeasurementProcessor` a more fundamental concept and symmetrically to `LogRecordProcessor` and `SpanProcessor`.

If not, then `Shutdown` and `ForceFlush` are not really needed, making `MeasurementProcessor` a slightly more capable version of Java's experimental `AttributesProcessor`.

https://github.com/open-telemetry/opentelemetry-java/blob/d61a5c3b37cf149ab87d1a1e2ec05dc44af20998/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AttributesProcessor.java#L31

However, there is also another route where these processors become a more fundamental concept. We could provide built-in processors that facilitate views and aggregations and push data to exporters.

Reviewer: Due to the nature of preaggregation, metrics work differently than logs/traces. MetricReader is the one which has Collect/Shutdown and ForceFlush (only push exporters have ForceFlush).

Author: Outdated - [60adbd3](https://github.com/open-telemetry/opentelemetry-specification/pull/4318/commits/60adbd3265005cd8cbf4e897075182245d1097f6).

|     Instrument X |-----------------> MeasurementProcessor +------------> In-memory state |
|     Instrument Y +                 |                      |            |                 |
|   Meter B        |                 +----------------------+            +-----------------+
|     Instrument Z |
|     ...          |                 +----------------------+            +-----------------+
|     ...          | Measurements... |                      | Metrics... |                 |
|     ...          |-----------------> MeasurementProcessor +------------> In-memory state |
|     ...          |                 |                      |            |                 |
|     ...          |                 +----------------------+            +-----------------+
+------------------+
```

### MeasurementProcessor operations

#### OnMeasure

`OnMeasure` is called when a `Measurement` is recorded. This method is called synchronously on the thread that emitted the `Measurement`, therefore it SHOULD NOT block or throw exceptions.

**Parameters:**

* `measurement` - a [Measurement](./api.md#measurement) that was recorded

> * `context` - the resolved `Context` (the explicitly passed `Context` or the current `Contex`)

Reviewer: ```suggestion
* `context` - the resolved `Context` (the explicitly passed `Context` or the current `Context`)
```

**Returns:** Void

For a `MeasurementProcessor` registered directly on SDK `MeterProvider`, the `measurement` mutations MUST be visible in next registered processors.

> 

Reviewer: Do we allow the processor to "drop" the measurement (e.g. the processor decided that it doesn't want the measurement) or other operations beyond modifications on the value and attributes?

Reviewer 5: Related question (thus decided to put it here). Shouldn't the processor also be used when evaluating `Enabled`? See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#enable

Reviewer 2: > the measurement processor does not change the instrument definition, and the measurement does not include a unit. Are there really use-cases for modifying the value?

Providing this feature without the ability to do unit conversion or drop measurements would be a miss. Can solve the lack of knowledge about unit by providing the processor access to instrument metadata. I think it could make sense to allow measurements processors to be configurable at the view level, in which case we might also consider allowing views to modify the unit of the resulting stream. Users could then compose a view which: 1. Adds a processor for unit conversion. 2. Adjusts the resulting stream's unit.

Reviewer 7: @Blinkuu About view-attached processors: I am generally wary of making the metrics SDK more complex, and the idea of making measurement processors view-specific has me vaguely worried. One fear is that this will limit the potential for MeasurementProcessors to be optimized.

If there are multiple readers and multiple measurement processors, do we evaluate the chain of processors per reader or once? I would prefer once.

---

I fear we're letting implementation details into the specification, if we dictate the use of a "next" processor here. I don't think a next processor is required. A better API for the processor IMO would be to _return_ a measurement, so a signature like

```
type Processor interface {
  Process(context.Context, Measurement) (_ Measurement, valid bool)
}
```

In other words, the return value is an optional Measurement. I suggest the specification use pseudo-code such as the following, to answer the question raised in https://github.com/open-telemetry/opentelemetry-specification/pull/4318/files#r1915514480 about the order of operations, and this makes the drop-behavior clear:

```
func (sdki *sdkInstrument) onEvent(ctx context.Context, m Measurement) {
  for _, processor in sdki.processors() {
    if mr, valid := processor.Process(ctx, m); valid {
      m = mr
    } else {
      // measurement was dropped
      return
    }
  }
  // instrument-specific logic for final processed measurement `m`
  // ...
}
```

If there's an argument in favor of view-specific measurement processors, I would suggest surveying the implementors of the various metrics SDKs view mechanisms for their opinion. Otherwise, I think it's much simpler to explain to users what's happening with measurement processors: they literally change the events that enter the SDK and are seen by all metric readers alike.

---

@Blinkuu regarding an array of measurements. This is another question that could impact performance. Unless we need it, I don't think one measurement should be allowed to become >1 measurement, in other words. In the example I gave of translating an UpDownCounter into two Counters, one input event becomes one output event.

Author: I'm happy to revisit the API. I think the one you're proposing aligns more with my original idea but doesn't allow duplicate measurements.

My personal use case does not require duplicating measurements. Furthermore, I don't work with metrics so much to have a strong opinion on this subject.

> In the example I gave of translating an UpDownCounter into two Counters, one input event becomes one output event.

Could you expand on this? How would you implement the split of counters in a `MeasurementProcessor` that implements your API:

```go
type Processor interface {
  Process(context.Context, Measurement) (_ Measurement, valid bool)
}
```

Reviewer 7: I'll retract my comment above. To achieve split counters, IMO a natural way to proceed would be to let each processor know the next processor--to split a metric you will create multiple Measurement types and call the next processor multiple times.

A `MeasuremenetProcessor` may freely modify `measurement` for the duration of the `OnMeasure` call.

#### Shutdown

Shuts down the processor. Called when the SDK is shut down. This is an opportunity for the processor to do any cleanup required.

> `Shutdown` SHOULD be called only once for each`MeasurementProcessor` instance. After the call to `Shutdow`, subsequent calls to `OnMeasure` are not allowed. SDKs SHOULD ignore these calls gracefully, if possible.

Reviewer: ```suggestion
`Shutdown` SHOULD be called only once for each `MeasurementProcessor` instance. After the call to `Shutdown`, subsequent calls to `OnMeasure` are not allowed. SDKs SHOULD ignore these calls gracefully, if possible.
```

`Shutdown` SHOULD provide a way to let the caller know whether it succeeded, failed or timed out.

`Shutdown` MUST include the effects of `ForceFlush`.

`Shutdown` SHOULD complete or abort within some timeout. `Shutdown` can be implemented as a blocking API or an asynchronous API which notifies the caller via a callback or an event. OpenTelemetry SDK authors can decide if they want to make the shutdown timeout configurable.

> #### ForceFlush

Reviewer 7: I wonder if ForceFlush is required or even meaningful. I looked over our API and SDK specification, and I see no text explaining how a SDK should respond if there are metric events concurrent with flush and shutdown operations. Since I view MeasurementProcessor as a synchronous processor, all the same ambiguity applies. If a user is calling metric events while  flushing or shutting down, the measurement processor may or may not have completed.

Since the processor is synchronous, it should have no buffered state to flush, it should just be synchronous.

Reviewer: +1, related to https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r1868070642

Author: Agree on the simplification, removed these methods from the spec [(60adbd3)](https://github.com/open-telemetry/opentelemetry-specification/pull/4318/commits/60adbd3265005cd8cbf4e897075182245d1097f6).

This is a hint to ensure that any tasks associated with `Measurements` for which the `MeasurementProcessor` had already received events prior to the call to `ForceFlush` SHOULD be completed as soon as possible, preferably before returning from this method.

<!--  TODO: Should we mingle with the Exporter concept here? For metrics, the only thing we care is that Measuremenets are processed before aggregation happens   -->

> In particular, if any `MeasurementProcessor` has any associated exporter, it SHOULD try to call the exporter's `Export` with all `Measurements` for which this was not already done and then invoke `ForceFlush` on it. If a timeout is specified (see below), the `MeasurementProcessor` MUST prioritize honoring the timeout over finishing all calls. It MAY skip or abort some or all `Export` or `ForceFlush` calls it has made to achieve this goal.

Reviewer: This is going to be tricky, what about MetricReader and the pull-based exporter? (e.g. Prometheus Exporter)

Author: Agreed, hence my comment. I referenced the `Exporter` here, but in the context of the Metrics SDK, it only makes sense to connect `MeasurementProcessor` and `MetricReader`.

To make it work, the `MeasurementProcessor` pipeline needs to exist between calls that record a new measurement, aka. `Add()` and `Record()` APIs defined on the `Instrument`, but before the actual aggregation happens, as at that point, we lose reference to `Attributes`.

So something along the lines of:

`Instrument ---Record()/Add()---> MeasurementProcessor(s) ---> Views -> Aggregations -> MetricReader`

However, I'm still trying to figure out if Views and Aggregations could somehow become built-in processors and whether it would even make sense.

**Parameters:**

* `context` - the resolved `Context` (the explicitly passed `Context` or the current `Context`)
* `measurement` - a [Measurement](./api.md#measurement) that was recorded
* `next` - this allows the `MeasurementProcessor` to pass the measurements to the next `MeasurementProcessor` in the chain. It can be a reference to the next `MeasurementProcessor`, a bound callback to invoke `OnMeasure` on the next processor in the chain without an explicit reference to the next processor, or something else. [OpenTelemetry SDK](../overview.md#sdk) authors MAY decide the language idiomatic approach.

**Returns:** Void

For a `MeasurementProcessor` registered directly on SDK `MeterProvider`, the `measurement` mutations MUST be visible in next registered processors.

A `MeasurementProcessor` MAY freely modify `measurement` for the duration of the `OnMeasure` call.

A `MeasurementProcessor` SHOULD invoke `next`. A `MeasurementProcessor` MAY decide to drop the `Measurement` by not invoking the next processor.

### Built-in processors

The standard OpenTelemetry SDK MUST implement default processor as described below.

#### DefaultProcessor

> This is an implementation of `MeasurementProcessor` which calculates an in-memory state from incoming `Measurements`.

Author: Specified in [756de83](https://github.com/open-telemetry/opentelemetry-specification/pull/4318/commits/756de836c6128ce340263184ea215d0b31ce359d). I used the `SHOULD` / `SHOULD NOT`, as in some languages it's impossible to make types private by design.

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## MeasurementProcessor

**Status**: [Development](../document-status.md)

`MeasurementProcessor` is an interface which allows hooks when a `Measurement` is recorded by an `Instrument`.

Built-in measurement processors are responsible for [Measurement Processing](#measurement-processing).

> `MeasurementProcessors` can be registered directly on SDK `MeterProvider` and they are invoked in the same order as they were registered.

Reviewer 4: My suggestion is to use the same wording as we use for Provider configuration in general.
i.e It MUST be possible to add measurement_processor as part of MeterProvider creation.
Implementations MAY allow modifications *after* MeterProvider is created. If they do, then all existing meter/instruments should reflect the effect of this.

> `MeasurementProcessors` can be registered directly on SDK `MeterProvider` and they are invoked in the same order as they were registered.

Author: I used the exact same wording we use for [LogRecordProcessor](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#logrecordprocessor) and [SpanProcessor](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-processor).

I think it should be consistent, it's the same concept.

> Each processor registered on the `MeterProvider` is part of a pipeline.

Author: Yes, it was considered. It looks similar, but in my opinion, Views are meant to work on a higher level of abstraction, not with individual measurements. The existing design is also deeply rooted in analysis of existing SDK implementations, where incorporating this concept into Views would be extremely difficult.

> Each processor registered on the `MeterProvider` is part of a pipeline.

Reviewer 8: Can you explain further?  Attribute filters from views are applied very early in the SDK's handling of a measurement.  They apply even [prior to cardinality limit enforcement](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#cardinality-limits). Implementing both of those features post-aggregation (i.e. not on measurements) would defeat one of their primary purpose, which is to limit memory consumption of the SDK, so I would be very surprised if they were not applied to individual measurements.

> Each processor registered on the `MeterProvider` is part of a pipeline.

Author: I'm not an expert on the Metrics SDKs, but after looking into the code when preparing this PR I concluded that supporting all the use cases we envisioned would require major overhaul of how Views work.

Please see https://github.com/open-telemetry/opentelemetry-specification/issues/4298#issuecomment-2486438111

Perhaps I'm wrong, and there's a neat way to make Views fully programmable. 

Regarding performance - MeasurementProcessor, like LogRecordProcessor and SpanProcessor is an opt-in feature.  If it's not used, the cost is a single if-statement, exactly the same as with https://github.com/MrAlias/opentelemetry-go/pull/1092/files#diff-2c6f5806e27a5ab81eeba6f9d6fff1a80ac35188c8e22deea1c97a77fdb04634

SDK MUST allow users to implement and configure custom processors.

SDK MUST ensure that the pipeline concludes with the built-in [DefaultProcessor](#defaultprocessor).

> The following diagram shows `MeasurementProcessor`'s relationship to other components in the SDK:

Author: Views are meant to be applied after Measurement Processors. They operate on a higher level of abstraction.

> The following diagram shows `MeasurementProcessor`'s relationship to other components in the SDK:

Reviewer 8: I realized cardinality limits should probably also be included, but we can wait until the [discussion below](https://github.com/open-telemetry/opentelemetry-specification/pull/4318#discussion_r2216274649) is resolved.
